core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java

```diff
@@ -96,27 +112,85 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeBoolean(this.isPrimary());
         out.writeOptionalString(this.getAssignedNodeId());
         out.writeOptionalWriteable(this.getUnassignedInfo());
+        out.writeVLong(remainingDelayNanos);
+        out.writeVInt(activeAllocationIds.size());
+        for (String id : activeAllocationIds) {
+            out.writeString(id);
+        }
 
-        Map<DiscoveryNode, Decision> ntd = this.getNodeDecisions();
-        out.writeVInt(ntd.size());
-        for (Map.Entry<DiscoveryNode, Decision> entry : ntd.entrySet()) {
-            entry.getKey().writeTo(out);
-            Decision.writeTo(entry.getValue(), out);
+        out.writeVInt(this.nodeExplanations.size());
+        for (NodeExplanation explanation : this.nodeExplanations.values()) {
+            explanation.writeTo(out);
         }
-        Map<DiscoveryNode, Float> ntw = this.getNodeWeights();
-        out.writeVInt(ntw.size());
-        for (Map.Entry<DiscoveryNode, Float> entry : ntw.entrySet()) {
-            entry.getKey().writeTo(out);
-            out.writeFloat(entry.getValue());
+    }
+
+    private NodeExplanation calculateNodeExplanation(DiscoveryNode node,
+                                                     Decision nodeDecision,
+                                                     Float nodeWeight,
+                                                     IndicesShardStoresResponse.StoreStatus storeStatus,
+                                                     String assignedNodeId,
+                                                     Set<String> activeAllocationIds) {
+        FinalDecision finalDecision;
+        StoreCopy storeCopy;
+        String finalExplanation;
+
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
+        } else if (nodeDecision.type() == Decision.Type.NO) {
+            finalDecision = FinalDecision.NO;
+            finalExplanation = "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision";
+        } else {
+            finalDecision = FinalDecision.YES;
+            finalExplanation = "the shard can be assigned";
         }
-        out.writeVLong(remainingDelayNanos);
+
+        if (storeStatus != null) {
+            final Throwable storeErr = storeStatus.getStoreException();
+            if (storeErr != null) {
+                finalDecision = FinalDecision.NO;
```

Comment:
The store has no impact on the decision whether a replica shard can be allocated to a node. This means that if the store for the existing data is corrupt we will still allocate a replica to that node (and wipe the folder). For a primary, it only plays a role if we require data to be there (`ShardRouting.allocatedPostIndexCreate() == false`) and the data is STALE.